framework/src/vectorpostprocessors/ElementVariablesDifferenceMax.C

```diff
@@ -0,0 +1,123 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2017 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "ElementVariablesDifferenceMax.h"
+
+// MOOSE includes
+#include "MooseVariable.h"
+
+// libMesh includes
+#include "libmesh/quadrature.h"
+
+template <>
+InputParameters
+validParams<ElementVariablesDifferenceMax>()
+{
+  InputParameters params = validParams<ElementVectorPostprocessor>();
+
+  params.addRequiredCoupledVar(
+      "compare_a",
+      "The first variable to evaluate the difference with, performed as \"compare_a - compare_b\"");
+  params.addRequiredCoupledVar("compare_b",
+                               "The second variable to evaluate the difference with, "
+                               "performed as \"compare_a - compare_b\"");
+
+  params.addParam<bool>(
+      "furthest_from_zero", false, "Find the difference with the highest absolute value");
+
+  return params;
+}
+
+ElementVariablesDifferenceMax::ElementVariablesDifferenceMax(const InputParameters & parameters)
+  : ElementVectorPostprocessor(parameters),
+    _a(coupledValue("compare_a")),
+    _b(coupledValue("compare_b")),
+    _a_value(declareVector(getVar("compare_a", 0)->name())),
+    _b_value(declareVector(getVar("compare_b", 0)->name())),
+    _max_difference(declareVector("Difference")),
+    _position_x(declareVector("X")),
+    _position_y(declareVector("Y")),
+    _position_z(declareVector("Z")),
+    _furthest_from_zero(getParam<bool>("furthest_from_zero"))
+{
+  // These are all single-value arrays
+  _a_value.resize(1);
+  _b_value.resize(1);
+  _max_difference.resize(1);
+  _position_x.resize(1);
+  _position_y.resize(1);
+  _position_z.resize(1);
+}
+
+void
+ElementVariablesDifferenceMax::execute()
+{
+  for (unsigned int qp = 0; qp < _qrule->n_points(); ++qp)
+  {
+    // Get the difference
+    const Real difference = _furthest_from_zero ? std::abs(_a[qp] - _b[qp]) : _a[qp] - _b[qp];
```

Comment:
I'm trying but still not understanding. This scenario almost sounds like the maximum values of `_a` and `_b` are being gathered based on themselves, not `_max_difference`.

`_a` and `_b` are updated at the same time, and only when a maximum difference is found at that particular location. In the gather phase they are being collected based on the global maximum value of `_max_difference`, not their individual values.

This revision is actually based on an earlier conversation with @dschwen, when it became apparent that the maximum difference did not need to be included in the MPI communications. It could simply be calculated from the values of the two fields in `finalize()`, thus allowing a reduction in the number of MPI calls being performed.